src/PhpDoc/TypeNodeResolver.php

```diff
@@ -361,9 +366,36 @@ private function resolveIdentifierTypeNode(IdentifierTypeNode $typeNode, NameSco
 			return new ErrorType();
 		}
 
+		if (!$this->mightBeConstant($typeNode->name) || $this->getReflectionProvider()->hasClass($stringName)) {
```

Comment:
I think this should just be `if ($this->getReflectionProvider()->hasClass($stringName)) {`.

Because if something looks like a constant `FOO_BAR` but there's a class `FOO_BAR`, I want `ObjectType` of the class, not the constant.

Reply:
This should still always result in the class taking precedence, but does a slightly cheaper heuristic check.

It's covered in the test as well: https://github.com/phpstan/phpstan-src/pull/1163/files#diff-f60365c259ff216f4d228e0c46729817e679144be2a7bf2e7395ace7fd344284R47

Reply:
I don't think it hits that condition:

```php
	private function mightBeConstant(string $name): bool
	{
		return preg_match('((?:^|\\\\)[A-Z_][A-Z0-9_]*$)', $name) > 0;
	}
```

This just matches UPPER_CASE names, right? But the constant in question has CamelCaps namespace.

Because when I read that code in case of `FOO_BAR` the first `return new ObjectType($stringName);` is skipped because `mightBeConstant() === true`. And then `tryResolveConstant` is successful because the constant exists. So constant takes precedence over class.

Reply:
The namespace isn't considered, only the last part of `$name` (so `BAR` in that case). If the constant took precedence the type in the test would be `10`, instead of the class name (`ConstantPhpdocType\BAR`).

Reply:
Now it makes sense to me :) I'm ready to merge this, thanks.